Sources/XMTPiOS/ApiClient.swift

```diff
@@ -0,0 +1,174 @@
+		}
+	}
+
+	static func envToUrl(env: XMTPEnvironment) -> String {
+		switch env {
+		case XMTPEnvironment.local: return "http://localhost:5556"
+		case XMTPEnvironment.dev: return "https://dev.xmtp.network:5556"
+		case XMTPEnvironment.production: return "https://production.xmtp.network:5556"
+		}
+	}
+
+	func setAuthToken(_ token: String) {
+		authToken = token
+	}
+
+	func batchQuery(request: BatchQueryRequest) async throws -> BatchQueryResponse {
+		do {
+			return try await rustClient.batchQuery(req: request.toFFI).fromFFI
+		} catch {
+			throw ApiClientError.batchQueryError(error.localizedDescription)
+		}
+	}
+
+	func query(request: QueryRequest) async throws -> QueryResponse {
+		do {
+			return try await rustClient.query(request: request.toFFI).fromFFI
+		} catch {
+			throw ApiClientError.queryError(error.localizedDescription)
+		}
+	}
+
+	func query(topic: String, pagination: Pagination? = nil, cursor: Cursor? = nil) async throws -> QueryResponse {
+		return try await query(request: makeQueryRequest(topic: topic, pagination: pagination, cursor: cursor))
+	}
+
+	func query(topic: Topic, pagination: Pagination? = nil) async throws -> QueryResponse {
+		return try await query(request: makeQueryRequest(topic: topic.description, pagination: pagination))
+	}
+
+	func envelopes(topic: String, pagination: Pagination? = nil) async throws -> [Envelope] {
+		var envelopes: [Envelope] = []
+		var hasNextPage = true
+		var cursor: Xmtp_MessageApi_V1_Cursor?
+
+		while hasNextPage {
+			let response = try await query(topic: topic, pagination: pagination, cursor: cursor)
+
+			envelopes.append(contentsOf: response.envelopes)
+
+			cursor = response.pagingInfo.cursor
+			hasNextPage = !response.envelopes.isEmpty && response.pagingInfo.hasCursor
+
+			if let limit = pagination?.limit, envelopes.count >= limit {
+				envelopes = Array(envelopes.prefix(limit))
+				break
+			}
+		}
+
+		return envelopes
+	}
+
+	func subscribe(topics: [String]) -> AsyncThrowingStream<Envelope, Error> {
+		return AsyncThrowingStream { _ in
+			Task {
+				let request = SubscribeRequest.with { $0.contentTopics = topics }
+				let req = try request.serializedData()
+				do {
+					// TODO: FIXME
+					//                    let subscription = try await self.rustClient.subscribe(req)
+					//                    // Run a continuous for loop polling and sleeping for a bit each loop.
+					//                    while true {
+					//                        let buf = try subscription.get_envelopes_as_query_response()
+					//                        // Note: it uses QueryResponse as a convenient envelopes wrapper.
+					//                        let res = try QueryResponse(serializedData: Data(buf))
+					//                        for envelope in res.envelopes {
+					//                            continuation.yield(envelope)
+					//                        }
+					//                        try await Task.sleep(nanoseconds: 50_000_000) // 50ms
+					//                    }
```

Comment:
Was this not implemented before? Just want to make sure we're not breaking subscribe.

Reply:
Oh it was. I should get this working before merging.

Reply:
I can help with the Rust side if needed